pubky-homeserver/src/core/layers/pubky_host.rs

```diff
@@ -40,25 +37,44 @@ where
     type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
 
     fn poll_ready(&mut self, cx: &mut std::task::Context<'_>) -> Poll<Result<(), Self::Error>> {
-        self.inner.poll_ready(cx).map_err(|_| unreachable!()) // `Infallible` conversion
+        self.inner.poll_ready(cx).map_err(|_| unreachable!())
     }
 
-    fn call(&mut self, req: Request<Body>) -> Self::Future {
+    fn call(&mut self, mut req: Request<Body>) -> Self::Future {
         let mut inner = self.inner.clone();
-        let mut req = req;
 
-        Box::pin(async move {
-            let headers_to_check = ["host", "pubky-host"];
+        // Use helper function to extract public key from headers or query parameter.
+        if let Some(public_key) = extract_pubky(&req) {
+            req.extensions_mut().insert(PubkyHost(public_key));
+        }
 
-            for header in headers_to_check {
-                if let Some(Ok(pubky_host)) = req.headers().get(header).map(|h| h.to_str()) {
-                    if let Ok(public_key) = PublicKey::try_from(pubky_host) {
-                        req.extensions_mut().insert(PubkyHost(public_key));
-                    }
+        Box::pin(async move { inner.call(req).await.map_err(|_| unreachable!()) })
+    }
+}
+
+/// Helper function to extract the public key from request headers or query parameter.
+fn extract_pubky(req: &Request<Body>) -> Option<PublicKey> {
+    // Check headers "host" and "pubky-host"
+    for header in ["host", "pubky-host"] {
+        if let Some(value) = req.headers().get(header) {
+            if let Ok(s) = value.to_str() {
+                if let Ok(pubky) = PublicKey::try_from(s) {
+                    return Some(pubky);
                 }
             }
+        }
+    }
 
-            inner.call(req).await.map_err(|_| unreachable!())
+    // Fallback: check query string for "pubky-host"
+    req.uri().query().and_then(|query| {
+        query.split('&').find_map(|pair| {
+            let mut parts = pair.splitn(2, '=');
+            if let (Some(key), Some(value)) = (parts.next(), parts.next()) {
+                if key == "pubky-host" {
```

Comment:
```suggestion
                if (key == "pubky-host" || key == "host") {
```
Simply by analogy with this one https://github.com/pubky/pubky-core/pull/94/files#diff-9fd5ac0adee2ef08e643ab5ed36387677497e1c670bad561673564fa5c208363R58

Reply:
Are you sure? I think in query param we should settle for one. I believe the original intention of the header check is to fallback to `host` (because browsers always attach this header) if the client user does not set a `pubky-host` header. I think this is makes sense because `host` will always be corrct and convertible to a `PublicKey` if the user is visiting himself on the browser (`https://{pubky_id}` or `pubky://{pubky_id}`). But when it comes to query params, this is an intentional decision on how you construct the URL and you either pick the convention to use `pubky-host` or `host` but having both does not add any benefit.

Reply:
I did not mean that I need it, just thought that we might need make behavior identical. Thanks for clarifying, makes sense